qualang_tools/utils/qua_utils.py

```diff
@@ -0,0 +1,237 @@
+    :param var_type: type of the a variable, can be either 'int' or 'fixed'.
+    :return: QUA for_ loop parameters (var, init, cond, update) as defined in https://qm-docs.qualang.io/api_references/qua/dsl_main?highlight=for_#qm.qua._dsl.for_.
+    """
+    assert len(array) > 0, "The array must be of length >0"
+
+    if np.isclose(np.std(np.diff(array)), 0):
+        increment = "lin"
+    elif np.isclose(np.std(array[1:] / array[:-1]), 0):
+        increment = "log"
+    else:
+        raise ValueError(
+            "The spacing of the input array must be linear or logarithmic."
+        )
+
+    start = array[0]
+    stop = array[-1]
+    if increment == "lin":
+        step = array[1] - array[0]
+
+        if var_type == "int":
+            assert float(
+                step
+            ).is_integer(), (
+                "When looping over a QUA int variable, the step must be integer."
+            )
+            if step > 0:
+                return var, int(start), var <= int(stop), var + int(step)
+            else:
+                return var, int(start), var >= int(stop), var + int(step)
+        elif var_type == "fixed":
+            assert (
+                np.abs(start) <= 8 and np.abs(stop) < 8
+            ), "fixed numbers are bounded to [8, 8)"
+            if step > 0:
+                return (
+                    var,
+                    float(start),
+                    var < float(stop) + float(step) / 2,
+                    var + float(step),
+                )
+            else:
+                return (
+                    var,
+                    float(start),
+                    var > float(stop) + float(step) / 2,
+                    var + float(step),
+                )
+    else:
+        step = array[1] / array[0]
+        assert (
+            np.abs(start) <= 8 and np.abs(stop) < 8
+        ), "fixed numbers are bounded to [8, 8)"
+        if step > 1:
+            return (
+                var,
+                float(start),
+                var < float(stop) * np.sqrt(float(step)),
+                var * float(step),
+            )
+        else:
+            return (
+                var,
+                float(start),
+                var > float(stop) * np.sqrt(float(step)),
+                var * float(step),
+            )
+
+
+def qua_arange(var, start, stop, step):
+    """
+     Function parametrizing the QUA `for_` loop from the numpy.arange() syntax.
+    :param var: the QUA variable that will be looped over (int or fixed).
+    :param start: Start of interval. The interval includes this value. Must be a Python variable.
+    :param stop: End of interval. The interval does not include this value. Must be a Python variable.
+    :param step: Spacing between values. Must be a Python variable.
+    :return: QUA for_ loop parameters (var, init, cond, update) as defined in https://qm-docs.qualang.io/api_references/qua/dsl_main?highlight=for_#qm.qua._dsl.for_.
+    """
+

```

Comment:
need to add a validation that:
var is qua variable
start, stop and step are not

Reply:
Done!